source/material_model/viscoelastic.cc

```diff
@@ -0,0 +1,700 @@
+
+              // Recalculate average values of viscosity, elastic shear modulus and viscoelastic (effective) viscosity
+              const std::vector<double> composition = in.composition[i];
+
+              // Average viscosity
+              const double average_viscosity = calculate_average_viscosity(composition,
+                                                                           viscosities,
+                                                                           viscosity_averaging);
+              // Average elastic shear modulus
+              const double average_elastic_shear_modulus =
+                calculate_average_elastic_shear_modulus(composition,
+                                                        elastic_shear_moduli,
+                                                        viscosity_averaging);
+
+              // Average viscoelastic viscosity
+              const double average_viscoelastic_viscosity =
+                calculate_average_viscoelastic_viscosity(average_viscosity,
+                                                         average_elastic_shear_modulus,
+                                                         dte);
+
+              // Calculate the current (new) viscoelastic stress, which is a function of the material
+              // properties (viscoelastic viscosity, shear modulus), elastic time step size, strain rate,
+              // vorticity and prior (inherited) viscoelastic stresses (see equation 29 in Moresi et al.,
+              // 2003, J. Comp. Phys.)
+              SymmetricTensor<2,dim> stress_new = ( 2. * average_viscoelastic_viscosity * deviator(in.strain_rate[i]) ) +
+                                                  ( ( average_viscoelastic_viscosity / ( average_elastic_shear_modulus * dte ) ) * stress_old ) +
+                                                  ( ( average_viscoelastic_viscosity / average_elastic_shear_modulus ) *
+                                                    ( symmetrize(rotation * Tensor<2,dim>(stress_old) ) - symmetrize(Tensor<2,dim>(stress_old) * rotation) ) );
+
+              // Stress averaging scheme to account for difference betweed fixed elastic time step
+              // and numerical time step (see equation 32 in Moresi et al., 2003, J. Comp. Phys.)
+              const double dt = this->get_timestep();
+              if (use_fixed_elastic_time_step == true && use_stress_averaging == true)
+                {
+                  stress_new = ( ( 1. - ( dt / dte ) ) * stress_old ) + ( ( dt / dte ) * stress_new ) ;
+                }
+
+              // Fill reaction terms
+              for (unsigned int j = 0; j < SymmetricTensor<2,dim>::n_independent_components ; ++j)
+                out.reaction_terms[i][j] = -stress_old[SymmetricTensor<2,dim>::unrolled_to_component_indices(j)]
+                                           + stress_new[SymmetricTensor<2,dim>::unrolled_to_component_indices(j)];
```

Comment:
This will not work with the iterated IMPES scheme. Can you add an Assert in Parse_parameters that makes sure we do not use the material model with iterated IMPES.

Reply:
Done.